srtcore/list.cpp

```diff
@@ -122,6 +122,17 @@ int CSndLossList::insert(int32_t seqno1, int32_t seqno2)
     const int offset  = CSeqNo::seqoff(m_caSeq[m_iHead].seqstart, seqno1);
     int       loc     = (m_iHead + offset + m_iSize) % m_iSize;
 
+    if (loc < 0)
+    {
+        // The size of the CSndLossList should be at least the size of the flow window.
+        // It means that all the packets sender has sent should fit within m_iSize.
+        // If the new loss does not fit, there is some error.
+        LOGC(mglog.Error, log << "IPE: New loss record is too old. Ignoring. "
```

Comment:
Not sure if this situation can be described as IPE here.

Reply:
BTW. Does it mean that `offset` is allowed to be negative in general, as long as it fits in the range from the current head and size?

Reply:
Potentially the following can happen:
1. Two packets with sequence numbers `A` and `B` (`A <% B`) are reported as lost.
2. Packet `A` is retransmitted and removed from the loss list.
3. Packet `A` is reported lost again, so that `CSeqNo::seqoff(m_caSeq[m_iHead].seqstart = B, seqno1 = A) < 0`.

So negative offset is allowed, it just has to fit into the list.

Reply:
Isn't however head moved forward after getting ACK? Isn't then a sequence that is earlier than head a sequence before the ACK? If this is true, then I can only imagine such a situation if reordering happened and the ACK that confirms the packet in question comes earlier than the lossreport that reports this sequence as lost, although it's actually outdated.

If this is the case, however, then it should be first checked if the upper sequence is below the head, and only if so should the loss record be rejected as a whole. Otherwise at least the range between the head and the upper sequence be recorded.

Reply:
The head is moved forward after the loss record is extracted and the corresponding packet is retransmitted.

If the same packet is reported lost again, it is placed before the head (negative `offset`), and the head position is updated.
```c++
// offset < 0
const int offset  = CSeqNo::seqoff(m_caSeq[m_iHead].seqstart, seqno1);
```
As I wrote above, a negative offset is allowed, it just has to fit into the list.

The `loc` identifies the position in the list to insert at. It is **not allowed** to be negative, otherwise it is out of bounds memory access.
```c++
int loc = (m_iHead + offset + m_iSize) % m_iSize;
```

Reply:
Ok, I understand that negative loc within the bounds is allowed, but I don't think it's correct to reject the whole report because of that. It should reject it only if the upper bound is behind the `seqstart`, and if it's not, the loss record should be accepted with the lower bound adjusted.

Reply:
Added check for upper sequence.